Federation rewrite with diaspora_federation gem

SuperTux88 opened the pull request:
This is the last step of #5114. I rewrote the whole incoming and outgoing federation code to use the new [federation-gem](https://github.com/SuperTux88/diaspora_federation).

This PR currently uses the latest commit on the develop-branch of the federation gem. I would release the 0.1.0 version  and update the PR after this PR was reviewed and tested successfully. Also I suggest to move the gem to the diaspora-organization after this PR is merged.

This PR also includes a complete rewrite of the notification-system, to fix the "sometimes no, sometimes one and sometimes two notification mails"-problem.

The following issues are fixed with this PR: #4727, #4697, #3666, #6579 (and many more bugs, where I didn't find an issue)

I think that #3519 also can be closed after this is merged, because this code doesn't exist anymore.

Note: The migration to cleanup participation (04c8329f7c895a3c132134f2e23a381439b9411c) can cause problems on big mysql databases if they have a `innodb_data_file_path` with `max:`. But this is not enabled by default on mysql (but somehow was on my local test-database). Should we add a note to the Changelog/Update-Notes for that? (see: http://stackoverflow.com/questions/21499169/innodb-delete-table-is-full)

Maybe someone can test the migrations with a bigger postgresql database? I only have a small test-DB (so the syntax should be OK), but I don't know if they cause trouble on bigger DBs (edge cases on old installations). I tested the migrations on a backup of my production mysql-database, so mysql should be OK.

I didn't write migrations to cleanup the `author_signature` of `Like`, `Comment` and `PollParticipation`, because @CSammy wants to make more database-optimizations and said that he can cleanup this then. Also the `author_signature` column of `Message` can be dropped completely after https://github.com/SuperTux88/diaspora_federation/issues/36 is done, so I didn't touch it now.

@cmrd-senya: maybe you can test this PR with your federation-tests?

---

(This part is not part of the initial PR text and gets abused as a ToDo list.)

ToDos before merging:
- [x] Move `SuperTux88/diaspora_federation` to `diaspora/diaspora_federation`
- [x] Verify migration compatibility on PostgreSQL
- [x] Manual testing of all the things
- [x] Second round of `diaspora_federation` gem reviewing
- [x] Release v0.1.0 of `diaspora_federation` gem

ToDos after merging:
- [ ] Work through all federation related issues, close them or move them into the federation repository if applicable
- [ ] Work through all notification related issues and see if those are fixed
- [ ] Create metabug for tracking database optimizations (actually, a milestone and single issues would work even better)
- [ ] Add a note regarding "full" MySQL tables into the upgrade notes.

SuperTux88 committed:
bump diaspora_federation to v0.1.0

SuperTux88 committed:
add more "diaspora to federation entity" methods

move it to Diaspora::Federation::Entities and use it in some tests, but most of the tests can be removed later.

SuperTux88 committed:
add new send workers

SuperTux88 committed:
fetch public key for unknown person

and do some refactorings

SuperTux88 committed:
implement :update_pod callback

SuperTux88 committed:
refactor `fetch_person_url_to` callback

SuperTux88 committed:
fix author join for federation callbacks with Poll

SuperTux88 committed:
create fetch_related_entity callback and cleanup unused callbacks

SuperTux88 committed:
fix federation tests

SuperTux88 committed:
add related entity to relayables and retractions

SuperTux88 committed:
create new receive workers

SuperTux88 committed:
save entities on receive

first implementation of receive_entity callback

SuperTux88 committed:
handle receive contact

SuperTux88 committed:
handle nested entities

SuperTux88 committed:
add ReceiveLocal worker

SuperTux88 committed:
handle ignored author on receive

SuperTux88 committed:
create NotificationService: send notifications after receive

SuperTux88 committed:
cleanup notifications and fix tests

SuperTux88 committed:
update profile

SuperTux88 committed:
handle existing guids on receive

SuperTux88 committed:
handle photo updates

SuperTux88 committed:
delete duplicates and create unique indexes:

* conversations
* messages
* photos
* polls
* poll answers
* poll participations

fixes #4697

SuperTux88 committed:
refactor message notifications

all visibilities are already in the conversation, so we can send
all notifications at once.

SuperTux88 committed:
receive local

* Contact: auto-follow-back
* Shareable: create share visibilities

SuperTux88 committed:
handle retractions

SuperTux88 committed:
fix TODOs in federation specs

SuperTux88 committed:
write specs for :receive_entity and extend factories

SuperTux88 committed:
write receive specs and delete old unused specs

also refactored receive methods

SuperTux88 committed:
move #add_contact_to_aspect from User-model to user_methods.rb

SuperTux88 committed:
remove Request class and refactor Connecting

SuperTux88 committed:
don't force remove contact on block

this creates inconsistent states, if you remove the block in the future

SuperTux88 committed:
remove old Private and Public Receiver

SuperTux88 committed:
rewrite attack vector specs using new federation

SuperTux88 committed:
remove NotifyLocalUsers worker

SuperTux88 committed:
first structure of Diaspora::Federation::Dispatcher

SuperTux88 committed:
handle user services

SuperTux88 committed:
create new contact for local receive

SuperTux88 committed:
remove user parameter from subscribers method

SuperTux88 committed:
remove after_dispatch hook

handle photos in StatusMessageCreationService

SuperTux88 committed:
create public and private dispatcher

SuperTux88 committed:
create Diaspora::Federation::Entities.build

SuperTux88 committed:
send public messages

SuperTux88 committed:
send private message

SuperTux88 committed:
update poor_mans_webmock.rb to handle the new federation

SuperTux88 committed:
mock new send worker in tests

SuperTux88 committed:
deliver profile update with new dispatcher

SuperTux88 committed:
deliver to pubsubhubbub

SuperTux88 committed:
enable some skipped tests again

SuperTux88 committed:
add DeferredRetraction worker

SuperTux88 committed:
federate retractions

SuperTux88 committed:
Contact retraction

SuperTux88 committed:
dispatch retractions

SuperTux88 committed:
send retractions only to remote subscribers

SuperTux88 committed:
send retractions for public targets publicly

SuperTux88 committed:
deliver retractions to services

SuperTux88 committed:
relay retractions for relayables

SuperTux88 committed:
deliver public posts to relay

SuperTux88 committed:
merge Federated::Shareable to Shareable

also drop xml-code for Shareable

SuperTux88 committed:
add resharers and participants to subscribers on public posts

SuperTux88 committed:
write spec for Diaspora::Federation::Entities to replace xml specs

SuperTux88 committed:
remove xml specs

SuperTux88 committed:
remove roxml

SuperTux88 committed:
remove salmon code

SuperTux88 committed:
use Diaspora::Federation::Dispatcher everywhere

SuperTux88 committed:
remove HttpMulti and HydraWrapper

SuperTux88 committed:
remove post fetcher

SuperTux88 committed:
write specs for retractions

SuperTux88 committed:
remove old SignedRetraction and RelayableRetraction

SuperTux88 committed:
create ShareVisibilities with batch import

and delete old batch worker and receiver

SuperTux88 committed:
only posts do have reshares

SuperTux88 committed:
delete old postzord dispatcher

SuperTux88 committed:
write specs for dispatcher

SuperTux88 committed:
refactoring delete from services

SuperTux88 committed:
remove more unused code and cleanup

SuperTux88 committed:
add root author to reshare subscribers

SuperTux88 committed:
allow to dispatch to specific subscribers

SuperTux88 committed:
send the profile to remote persons on start shareing with them

SuperTux88 committed:
retract relayables if the author is ignored by the parent author

SuperTux88 committed:
relay relayables after receive

also refactored relayable specs to use `let`

SuperTux88 committed:
deliver message always to all participants

message doesn't include Diaspora::Relayable anymore

SuperTux88 committed:
participates root author when receiving a reshare

SuperTux88 committed:
create module for author association

SuperTux88 committed:
move Guid module to fields

SuperTux88 committed:
create module for target association

SuperTux88 committed:
Participation is not relayable anymore

only deliver to parent author

SuperTux88 committed:
move Federated::Generator to Diaspora::Federated::Generator

SuperTux88 committed:
disconnect when receiving a Contact with sharing=false

SuperTux88 committed:
don't participate own posts on creation

SuperTux88 committed:
add migration to cleanup participations

* remove author_signature column
* delete all self-participating participations
* delete all participations where author and post-author are remote
* delete all duplicate participations
* delete all participations without post
* create unique index

SuperTux88 committed:
send retractions for public relayables privately if the parent is remote

SuperTux88 committed:
don't send relayables back to sender pod

send retraction for relayable to target author if retracted by parent author

SuperTux88 committed:
add author to subscribers of public shareables to ensure local delivery

diaspora-code-review commented on app/views/notifications/_notification.haml:
```diff
@@ -1,9 +1,9 @@
   .unread-toggle.pull-right
     %i.entypo-eye{title: (note.unread ? t("notifications.index.mark_read") : t("notifications.index.mark_unread"))}
-  - if note.type == "Notifications::StartedSharing" && contact = current_user.contact_for(note.effective_target)
+  - if note.type == "Notifications::StartedSharing" && contact = current_user.contact_for(note.target)
```
Assignment in condition - you probably meant to use `==`.

denschub replied:
Ignore.

diaspora-code-review commented on app/workers/send_base.rb:
```diff
@@ -0,0 +1,28 @@
+  class SendBase < Base
+    sidekiq_options queue: :http, retry: 0
+
+    MAX_RETRIES = (rt = AppConfig.environment.sidekiq.retry.get) ? rt.to_i : 0
```
Assignment in condition - you probably meant to use `==`.

denschub replied:
`to_i` [returns 0 if an invalid string/nil is provided](http://ruby-doc.org/core-2.2.0/String.html#method-i-to_i):

```ruby
MAX_RETRIES = AppConfig.environment.sidekiq.retry.get.to_i
```

diaspora-code-review commented on lib/diaspora/federation/entities.rb:
```diff
@@ -0,0 +1,258 @@
+module Diaspora
+  module Federation
+    module Entities
+      def self.build(entity)
```
Cyclomatic complexity for build is too high. [14/6]

denschub replied:
Yes, @diaspora-code-review, you are perfectly right, the cyclomatic complexity is way too high for a normal function. But dude, take a look at the source! It's simply calling some function depending on the incoming objects, mate. That's not an issue! I am sure you can be happy with that, can't you?

Oh, wait, I am talking to a bot. It's going down.

SuperTux88 replied:
fixed :smiley_cat:

denschub commented on Gemfile:
```diff
@@ -136,7 +136,6 @@ gem "leaflet-rails",       "0.7.7"
 gem "nokogiri",          "1.6.8"
 gem "redcarpet",         "3.3.4"
 gem "twitter-text",      "1.13.4"
-gem "roxml",             "3.1.6"
```
🎉 🎈 🍰 ❤️

denschub commented on app/models/contact.rb:
```diff
@@ -93,26 +80,45 @@ def self.contact_contacts_for(user, person)
+    [person]
+  end
+
+  # creates or updates a contact with active sharing flag. Returns nil if allready sharing.
```
Nit: already.

denschub commented on app/models/contact.rb:
```diff
@@ -93,26 +80,45 @@ def self.contact_contacts_for(user, person)
+
+    return if contact.sharing
+
+    contact.tap do |contact|
```
You like `.tap`, don't you? :) However, I am not too sure if it adds anything here besides looking fancy. Not using tap and returning the object without tap actually is a line shorter and looks less complicated:

```ruby
contact.sharing = true
contact.save!
contact
```

Also, if you just update a single attribute,

```ruby
contact.update(sharing: true)
contact
```

Would work, too.

SuperTux88 replied:
yes, I think the solution with `update` is better, I'll change that.

SuperTux88 committed:
remove diaspora_handle setter to set author

denschub commented on config/defaults.yml:
```diff
@@ -187,7 +187,7 @@ defaults:
     outbound:
       send: false
-      handle: 'relay@relay.iliketoast.net'
+      url: 'https://relay.iliketoast.net/receive/public'
```
👍

denschub commented on config/initializers/diaspora_federation.rb:
```diff
@@ -63,64 +63,95 @@
+      when DiasporaFederation::Entities::Retraction
+        Diaspora::Federation::Receive.retraction(entity, recipient_id)
+      else
+        persisted = case entity
```
Note to myself: think about if/how we can make this prettier.

jhass replied:
A hash, moving it out of sight of the glue code and `public_send` is probably acceptable here.

```rb
Diaspora::Federation::Receive.perform(entity)

def self.perform(entity)
  receiver = ENTITY_RECEIVERS.fetch(entity.class) { raise ... }
  public_send(receiver, entity)
end

denschub commented on config/initializers/diaspora_federation.rb:
```diff
@@ -63,64 +63,95 @@
+        pod.status = :no_errors unless Pod.statuses[pod.status] == Pod.statuses[:version_failed]
+      else
+        pod.status = :http_failed
+        pod.error = "FederationError: HTTP-Status-Code was: #{status}"
```
Nit: english does not use hyphens to combine words
```diff
- HTTP-Status-Code
+ HTTP status code
```

denschub commented on features/support/poor_mans_webmock.rb:
```diff
@@ -9,8 +9,14 @@ def perform(_sender_atom_url)
-  class HttpMulti < Base
-    def perform(_user_id, _enc_object_xml, _person_ids, _retry_count=0)
+  class SendPrivate < Base
+    def perform(_sender_id, _obj_str, _targets, _retry_count=0)
```
What do you think about `def perform(*_args)`? This would reduce the amount of changes we need to do if we ever change the signature of this function and we have other specs to check if the functions are called correctly.

denschub commented on lib/diaspora/federation/receive.rb:
```diff
@@ -0,0 +1,322 @@
+
+      def self.contact(entity)
+        recipient = Person.find_by(diaspora_handle: entity.recipient).owner
+        if entity.sharing.to_s == "true"
```
Note to myself: No, @SuperTux88 is not crazy. The federation gem is only capable of passing strings at the moment and the conversion to a Boolean happens by AR. This is the only place where a check of that kind is needed, so the adjustment to the federation gem has a low priority, but will be done eventually.

denschub commented on lib/diaspora/taggable.rb:
```diff
@@ -1,3 +1,4 @@
+
```
Oh hi there little newline! You look pretty out of place here, but I will help you finding your original destination!

denschub commented on spec/workers/send_base_spec.rb:
```diff
@@ -0,0 +1,20 @@
+    expect(retry_delay).to be < 316
+  end
+
+  it "increases the retry time for every retry exponentially " do
```
Nit: this is not an exponential increase.

denschub commented:
So, before adding any more comments: **Thank you for your work**. I know that you spent over a year on the federation gem and on getting this first PR out. This is awesome and I guess it is fair to say that this is the biggest change for diaspora so far. It is very sad that, if everything goes well, no user will ever notice what was achieved here, but that's the fun of backend work, isn't it?

I finished a first round of review, mainly looking for obvious mistakes (which I did not find any) and style improvements. The migrations look fine to me, but I will run them on a copy of Geraspora's database (which, coincidentally, is a Postgres database) to make sure. Also, I do expect some more changes to come.

But... great work so far! ❤️

*Finally, a note to everyone else: I know everyone is excited, but please keep the comments of this PR rather clean. Please add your heart and party emoticons to the PR, but please do not post any comments unless needed. If you have questions about a change or you want to suggest something but you are not sure if this is something that should be done, please contact a core team member via IRC. I do expect some technical discussions about some implementations in this PR and everything will be easier if we reduce the amount of off-topic comments. Thanks! <3*

denschub commented:
As for answering the questions in the initial PR text as well as keeping track of open tasks, I've added a list to the PR description.

cmrd-senya commented:
> @cmrd-senya: maybe you can test this PR with your federation-tests?

Yes, I'll test this during the next week.

jhass commented on app/controllers/blocks_controller.rb:
```diff
@@ -35,9 +35,8 @@ def destroy
-      current_user.disconnect(contact, :force => true)
-    end
+    contact = current_user.contact_for(person)
+    current_user.disconnect(contact) if contact
```
Style question: Should start to make use of `try` for these kind of cases?

```rb
current_user.contact_for(person).try {|contact| current_user.disconnect(contact) }
```

jhass commented on app/models/conversation.rb:
```diff
@@ -59,15 +44,12 @@ def set_read(user)
-
   def participant_handles
-    self.participants.map{|p| p.diaspora_handle}.join(";")
+    participants.map(&:diaspora_handle).join(";")
```
How likely is it that we'll need the whole objects anyway if this is called? In other words could it be beneficial to `pluck` instead or are we going to load the whole objects anyway?

SuperTux88 replied:
this method is called while dispatching, and the participants are probably already loaded, because he calls `subscribers` before he calls `participant_handles`.

jhass commented on app/models/message.rb:
```diff
@@ -1,58 +1,22 @@
-    self.conversation = parent
+  def conversation_guid=(guid)
+    cnv = Conversation.find_by(guid: guid)
+    self.conversation = cnv if cnv
```
Similar micro optimization consideration, how likely is it that we'll need the whole `Conversation` object later if this was called? Else we could save the allocation with `super if Conversation.exists?`.

Silently ignoring the assignment feels worrying too, but probably out of scope here to fix. Might be worth a TODO comment nonetheless.

jhass commented on app/models/message.rb:
```diff
@@ -62,17 +26,23 @@ def increase_unread(user)
+    if author.local?
+      conversation.participants
+    else # for relaying, TODO: can be removed when messages are not relayed anymore
+      conversation.participants.select(&:remote?)
```
Same again, is the whole collection loaded anyway? If not can we turn this into a DB query?

jhass commented on app/models/notification.rb:
```diff
@@ -3,93 +3,56 @@
-    end
-    return_note.email_the_user(target, actor) if return_note
-    return_note
+    where(opts.merge!(recipient_id: recipient.id)).order("updated_at desc")
```
Nit: capitalize `DESC`.

jhass commented on app/models/notification.rb:
```diff
@@ -3,93 +3,56 @@
+    find_or_initialize_by(recipient: recipient, target: target, unread: true).tap do |notification|
+      notification.actors |= [actor]
+      # Explicitly touch the notification to update updated_at whenever new actor is inserted in notification.
+      notification.touch unless notification.new_record?
```
Micro: We shouldn't need to `save!` if the record has no changes, just `touch` and we shouldn't need to `touch` if the record has changes, `save!` will do it:
```rb
if notification.new_record? || notification.changed?
  notification.save!
else
  notification.touch
end

jhass commented on app/models/notifications/also_commented.rb:
```diff
@@ -1,17 +1,26 @@
+    def self.notify(comment, _recipient_user_ids)
+      actor = comment.author
+      commentable = comment.commentable
+      recipients = commentable.participations.map(&:author).select(&:local?) - [commentable.author, actor]
```
Same questions as before, do we need to force the collections here or could this be a `joins().where().not()`?

jhass commented on app/models/notifications/also_commented.rb:
```diff
@@ -1,17 +1,26 @@
-    Post.where(:id => self.target_id).first
+      recipients.each do |recipient|
+        notification = concatenate_or_create(recipient.owner, commentable, actor)
+        notification.email_the_user(comment, actor) if notification
```
Another `.try { }` candidate.

jhass commented on app/models/notifications/private_message.rb:
```diff
@@ -1,15 +1,29 @@
+      recipients.map(&:owner).each do |recipient|
+        message.increase_unread(recipient)
+        new(recipient: recipient).email_the_user(message, message.author)
+      end
```
Same story about loading the entire collections, I'll skip any further ones for now.

jhass commented on app/models/photo.rb:
```diff
@@ -81,13 +72,11 @@ def ownership_of_status_message
-  def self.diaspora_initialize(params = {})
-    photo = shareable_initialize(params)
+  def self.diaspora_initialize(params={})
+    photo = new(params.to_hash.stringify_keys.slice(*column_names, "author"))
```
`to_h`

jhass commented on app/models/poll_participation.rb:
```diff
@@ -1,43 +1,18 @@
-    self.author = Person.find_or_fetch_by_identifier(nh)
+  def poll_answer_guid=(new_poll_answer_guid)
+    self.poll_answer = PollAnswer.find_by(guid: new_poll_answer_guid)
   end
```
Same question as for `conversation_guid=`.

jhass commented on app/models/post.rb:
```diff
@@ -132,13 +129,13 @@ def like_for(user)
-  def mutable?
-    false
+  def self.diaspora_initialize(params)
+    new(params.to_hash.stringify_keys.slice(*column_names, "author"))
```
`to_h` vs `to_hash`.

jhass commented on app/models/post.rb:
```diff
@@ -152,4 +149,10 @@ def comment_email_subject
+
+  def subscribers
+    subs = super
+    subs += resharers + participants if public?
```
Do we own the object we get from the parent at this point? `Array.concat` could save a few array allocations here.

SuperTux88 replied:
> Do we own the object we get from the parent at this point?

Yes, the parent is only called if we own it.

jhass replied:
I mean does the parent hold a reference to it otherwise?

SuperTux88 replied:
Sorry, I misunderstood you, no the parent doesn't hold a reference to it.

jhass commented on app/models/user/connecting.rb:
```diff
@@ -2,56 +2,59 @@
-      contact.update_attributes(:sharing => false)
-    else
-      contact.update_attributes(:receiving => false)
+      disconnect_contact(contact, :receiving, !contact.sharing)
```
This call could benefit from keyword arguments:
```rb
disconnect_contact(contact, direction: :receiving, destroy: !contact.sharing)
``

jhass commented on app/services/notification_service.rb:
```diff
@@ -0,0 +1,21 @@
+    Message       => Notifications::PrivateMessage,
+    Reshare       => Notifications::Reshared,
+    Contact       => Notifications::StartedSharing
+  }.freeze
```
Always wrapping the values into an array and getting rid of the `[*...]` below is probably a bit cleaner here.

jhass commented on app/services/status_message_creation_service.rb:
```diff
@@ -42,19 +42,24 @@ def add_poll(status_message, params)
+      status_message.photos.each do |photo|
+        photo.public = status_message.public
+        photo.pending = false
+      end
```
Can this be an `update_all` or do we rely on the validations here?

jhass commented on app/workers/deferred_retraction.rb:
```diff
@@ -0,0 +1,18 @@
+      user = User.find(user_id)
+      subscribers = Person.where(id: recipient_ids)
+      object = Retraction.new(retraction_data.deep_symbolize_keys, subscribers)
+      opts = HashWithIndifferentAccess.new(opts)
```
Okay this probably a bit too fundamental at this point, but I would have loved for the lower level to not rely on HWIAs and expect normalized to string keys option hashes. Or given Ruby 2.3's Symbol GC we could even consider normalizing to symbols.

SuperTux88 replied:
I think we should normalize to symbols once diaspora officially supports ruby 2.3.

jhass commented on config/initializers/diaspora_federation.rb:
```diff
@@ -6,8 +6,8 @@
-    on :fetch_person_for_webfinger do |handle|
-      person = Person.find_local_by_diaspora_handle(handle)
+    on :fetch_person_for_webfinger do |diaspora_id|
+      person = Person.where(diaspora_handle: diaspora_id, closed_account: false).where.not(owner: nil).take
```
Why `take` over `first`? It's unlikely for this to ever match more than one record, but if for some reason it does, `first` would at least give a deterministic result, due ordering by PK.

jhass commented on lib/diaspora/federation/dispatcher.rb:
```diff
@@ -0,0 +1,84 @@
+      end
+
+      def self.defer_dispatch(sender, object, opts={})
+        object.save if object.new_record?
```
Why the explicit check, is there a reason why it shouldn't be saved if there are just changed attributes? Otherwise `save` is a noop if the object is `persisted? && !changed?`

jhass commented on lib/diaspora/federation/dispatcher.rb:
```diff
@@ -0,0 +1,84 @@
+      attr_reader :sender, :object, :opts
+
+      def self.object_should_be_processed_as_public?(object)
+        object.respond_to?(:public?) && object.public?
```
Style alternative: `object.try(:public?) || false`

jhass commented on lib/diaspora/federation/dispatcher.rb:
```diff
@@ -0,0 +1,84 @@
+      def self.object_should_be_processed_as_public?(object)
+        object.respond_to?(:public?) && object.public?
+      end
+      private_class_method :object_should_be_processed_as_public?
```
`def` returns the methods name as symbol now, so in theory you could write `private_class_method def self.`, not sure we want this though.

jhass commented on lib/diaspora/federation/dispatcher.rb:
```diff
@@ -0,0 +1,84 @@
+      end
+
+      def deliver_to_local(people)
+        obj = object.respond_to?(:object_to_receive) ? object.object_to_receive : object
```
`object.try(:object_to_receive) || object`

SuperTux88 replied:
this doesn't work, because `object_to_receive` can return nil, and then it shouldn't use `object`.

jhass replied:
The cleaner variant would be for all possible objects to implement `object_to_receive` and return `self`, but I have no idea if that's feasible here then.

SuperTux88 replied:
I can add a `object_to_receive` to `Diaspora::Federated::Base` then?

jhass replied:
Fine with me.

jhass commented on lib/diaspora/federation/dispatcher.rb:
```diff
@@ -0,0 +1,84 @@
+
+      def deliver_to_local(people)
+        obj = object.respond_to?(:object_to_receive) ? object.object_to_receive : object
+        Workers::ReceiveLocal.perform_async(obj.class.to_s, obj.id, people.map(&:owner_id)) if obj
```
The guard if feels a bit hidden here. To please rubocop about it I'd use a guard clause instead `return unless obj`

jhass commented on lib/diaspora/federation/dispatcher.rb:
```diff
@@ -0,0 +1,84 @@
+        end
+      end
+
+      def for_services
```
Just `each_service`?

jhass commented on lib/diaspora/federation/dispatcher/public.rb:
```diff
@@ -0,0 +1,46 @@
+        def additional_target_urls
+          [].tap do |urls|
+            urls << AppConfig.relay.outbound.url if AppConfig.relay.outbound.send? && object.instance_of?(StatusMessage)
+          end
```
Too smart IMO, just
```rb
return unless AppConfig.relay.outbound.send? || !object.instance_of(StatusMessage)
[AppConfig.relay.outbound.url]
```

jhass commented on lib/diaspora/federation/entities.rb:
```diff
@@ -0,0 +1,258 @@
+        when Profile then profile(entity)
+        when Reshare then reshare(entity)
+        when Retraction then build_retraction(entity)
+        when StatusMessage then status_message(entity)
```
Mh, so this is essentially the same list and mapping we have in the glue code earlier?

SuperTux88 replied:
same mapping, but other direction. This is outgoing (rails-model -> federation-entity), the other is incoming (federation-entity -> rails-model)

jhass commented on lib/diaspora/federation/entities.rb:
```diff
@@ -0,0 +1,258 @@
+
+      def self.post(post)
+        case post
+        when StatusMessage then status_message(post)
```
Style nit: I kinda dislike `then` personally and prefer newlines

jhass commented on lib/diaspora/federation/receive.rb:
```diff
@@ -0,0 +1,322 @@
+
+      def self.photo(entity)
+        author = author_of(entity)
+        persisted_photo = load_from_db(Photo, entity.guid, author)
```
Let's avoid acronyms as much as possible, `db` -> `database`.

jhass commented on lib/diaspora/federation/receive.rb:
```diff
@@ -0,0 +1,322 @@
+      end
+
+      def self.retraction(entity, recipient_id)
+        object = entity.target_type.constantize.where(guid: entity.target_guid).take!
```
Does the layer guarantee that the `constantize` can't fail here?

SuperTux88 replied:
Yes, the federation-gem already checks the target on receive, and checks if the sender of the retraction is allowed to delete this entity, otherwise it would raise a `DiasporaFederation::Entities::Retraction::TargetNotFound` or `DiasporaFederation::Federation::Receiver::InvalidSender`

jhass replied:
But `target_type` comes from the XML? Didn't we want to decouple the XML/entity names from the class names?

SuperTux88 replied:
But currently they are the same. Should I create a mapping for String <-> Class instead of `.to_s` and `constantize`?

jhass replied:
I would say so. This is no different to using `const_get` or similar where your big case/when's are currently.

jhass commented on lib/diaspora/shareable.rb:
```diff
@@ -29,50 +30,61 @@ def self.included(model)
+    def subscribers
+      user = author.owner
+      if public?
+        user.contact_people + [author]
```
`[*user.contact_people, author]` should avoid one array allocation here.

jhass commented on lib/diaspora/shareable.rb:
```diff
@@ -29,50 +30,61 @@ def self.included(model)
+      end
+
+      def owned_by_user(user)
+        user.person.send(table_name).where(pending: false)
```
`public_send`

jhass commented:
So far I couldn't spot a merge blocker, good work!

SuperTux88 committed:
add #object_to_receive to Diaspora::Federated::Base

SuperTux88 committed:
use private_class_method with returned symbol of def

SuperTux88 committed:
extract mappings between diaspora and the federation gem

SuperTux88 commented:
@denschub and @jhass thx for the review so far, I updated the PR with the changes you mentioned.

cmrd-senya commented:
I set up the test machine to test `origin/develop` against your branch. The testsuite is very limited ATM, and the little set I have now [passes](http://190.15.209.124:8010/builders/regression_testing/builds/21/steps/shell_5/logs/stdio) well.

I can dedicate this week to extending the testsuite. Then I can set the suite to test your branch against `origin/stable` and itself also (or any other version actually).

@SuperTux88, if you have any suggestions what features to test first, for example if you're unsure with something or some part is more complex and therefore requires more attention, tell me.

SuperTux88 commented:
Thanks. I think that we don't need many tests in this federation-tests, only the basics to test that federation itself work, and we can test the other cases better with normal unit-/integration-tests.

So if you have start/stop sharing, write/delete public/private post and write/delete comments (best with 3 pods, to test relayables), it would cover most of the federation.

I tested locally with 3 pods (one with this branch and two develop), and everything that didn't work properly was because of stuff still broken in develop :wink:

SuperTux88 committed:
remove participation parent author check

the check that the participation parent author is local is now done in
the diaspora_federation gem

diaspora-code-review commented on app/workers/receive_base.rb:
```diff
@@ -0,0 +1,33 @@
+    include Diaspora::Logging
+
+    # don't retry for errors that will fail again
+    def filter_errors_for_retry
```
Method has too many lines. [21/20]

SuperTux88 replied:
yes, I know, but after removing the deprecated errors, it will be under 20 lines again. I hope that is OK for you, holy @diaspora-code-review :innocent:

SuperTux88 committed:
the xml parameter is now filtered in the diaspora_federation gem

SuperTux88 committed:
use DiasporaFederation::Salmon::MagicEnvelope for federation debug logging

SuperTux88 committed:
don't participate own posts

SuperTux88 commented:
I did a few more improvements on this PR and on the gem.

SuperTux88 commented:
aaaaaand I broke something :scream_cat:

I finally found why a user participates his own posts ... so that the posts are shown in his "my activity"-stream. Funny is, that the tests for this are broken anyway, because there is [a test](https://github.com/diaspora/diaspora/blob/develop/features/mobile/activity_stream.feature#L12-L15) that a users own posts doesn't appear in the activity-stream (which they do in the current stable and develop version). Because of this broken test, I didn't notice this before I opened this PR ... only the second test fails now. So I need to fix the tests too.

But I need to sleep now, I'll fix this later :sleeping:

SuperTux88 committed:
add own posts to activity-stream again

cmrd-senya commented:
Intermediate test results.

I [extended](https://github.com/cmrd-senya/diaspora_federation_behavior/commit/d077a5f41f3a21334853b96187b7b2e3def92ff5) the testsuite with a test for conversation (post and comments and retractions) for two people on different pods. The test runs [properly](http://190.15.209.124:8010/builders/regression_testing/builds/26/steps/shell_5/logs/stdio) with this branch against `origin/develop`.

Now, I'll implement and conduct similar tests for 3 pods case to test relayables properly.

cmrd-senya commented:
Ok, I [implemeted](https://github.com/cmrd-senya/diaspora_federation_behavior/commit/50ae0d349ee07490d4a95cf846a1efecc5c4a3ce) [a test](https://github.com/cmrd-senya/diaspora_federation_behavior/commit/50ae0d349ee07490d4a95cf846a1efecc5c4a3ce#diff-16577f811afc79e70a514367c7880c9fR81) for 3 pods case conversation.

I set the test suite to run the following sets of branches:

pod1 | pod2 | pod3
------------ | ------------- | -------------
origin/develop | origin/stable | origin/stable
origin/develop | origin/stable | origin/develop
origin/develop | SuperTux88/federation-gem-salmon | origin/stable
origin/develop | SuperTux88/federation-gem-salmon | origin/develop

The following steps are taken in the test:
0) user@pod2 shares with user@pod1 and user@pod3
1) Post by user@pod2
2) Comment by user@pod1 and user@pod3
3) Comment retractions by user@pod2
4) Post retraction by user@pod2

Each step is verified to be reflected on every pod. The test is run using both public and private post method.

All the tests [have passed](http://190.15.209.124:8010/builders/regression_testing_3pods/builds/4/steps/shell_5/logs/stdio) with no failures.

The buildbot installation is set to track changes in origin/develop and SuperTux88/federation-gem-salmon branches. Once it notice some changes, it'll run tests automatically, and new builds will appear [here](http://190.15.209.124:8010/builders/regression_testing) for 2-pods tests and [here](http://190.15.209.124:8010/builders/regression_testing_3pods) for 3-pods tests. But I haven't tested that too well and I also can't promise that the test framework will work stable without my supervision.

Note that all the test I perform work in a single process mode of diaspora run (like in develop). If anyone thinks it's important to test it with sidekiq, I could change the setting and rerun the tests.

SuperTux88 commented:
Thanks!

Single process is ok, but you need at least two unicorn worker for working federation (one to send out the entity, and one to handle the incoming webfinger request if the other pod doesn't know the sender already). I don't know if that is the default, because I had deadlock-problems once because of this. If the pods webfinger the persons before they receive the first entity, it works also with one worker.

cmrd-senya commented:
I have `unicorn_worker` set to 4 on all pods. By single process mode I meant the `single_process_mode`  setting in `diaspora.yml`, which, I guess, only embeds jobs processing to unicorn, but still keeps workers count > 1.

denschub commented:
Migrations are working on PostgreSQL and a copy of Geraspora's production database. To my surprise, the migrated database is 29,55% smaller than the old one. Quite impressive given the fact that even more database optimizations will follow!

denschub commented:
Yihhhaaaa! 🎉 

I've merged this so I'm able to test this on Geraspora. I hope this is stable enough for a dev-merge, but we can always file follow-ups.

Thanks again for your work here! ❤️

Flaburgan commented:
>the migrated database is 29,55% smaller than the old one

Was that migration done from a master database or from a develop one with #6586 already applied?

Awesome news! Big up for the new federation!

denschub commented:
Yes, #6586 was already applied, so the size reduction for 0.5.0.0 -> 0.6.0.0 will be even more impressive.

Flaburgan commented:
I updated d-fr which now runs the last develop commits.
Before the pg_dump was 7972654124B
After it is 4846210865B
Here is the result of the `db:migrate` if you're interested about the time: https://pastebin.mozilla.org/8880142

Thank you very much!
